ocaml/libs/tracing/tracing.ml

```diff
@@ -382,123 +390,114 @@ module Spans = struct
-        | Some span_list ->
-            if List.length span_list < Atomic.get max_spans then
-              Hashtbl.replace spans key (span :: span_list)
-            else if not_throttled () then
-              debug "%s exceeded max traces when adding to span table"
-                __FUNCTION__
-    )
+    match SpanMap.find_opt key spans with
+    | None ->
+        if SpanMap.cardinal spans < Atomic.get max_traces then
+          SpanMap.add key [span] spans
+        else (
+          if not_throttled () then
+            debug "%s exceeded max traces when adding to span table"
+              __FUNCTION__ ;
+          spans
+        )
+    | Some span_list ->
+        if List.length span_list < Atomic.get max_spans then
+          SpanMap.add key (span :: span_list) spans
+        else (
+          if not_throttled () then
+            debug "%s exceeded max traces when adding to span table"
+              __FUNCTION__ ;
+          spans
+        )
 
-  let remove_from_spans span =
+  let add_to_spans ~span = update_spans add_to_spans_unlocked span
+
+  let remove_from_spans_unlocked spans span =
     let key = span.Span.context.trace_id in
-    Xapi_stdext_threads.Threadext.Mutex.execute lock (fun () ->
-        match Hashtbl.find_opt spans key with
-        | None ->
-            if not_throttled () then
-              debug "%s span does not exist or already finished" __FUNCTION__ ;
-            None
-        | Some span_list ->
-            ( match
-                List.filter (fun x -> x.Span.context <> span.context) span_list
-              with
-            | [] ->
-                Hashtbl.remove spans key
-            | filtered_list ->
-                Hashtbl.replace spans key filtered_list
-            ) ;
-            Some span
+    match SpanMap.find_opt key spans with
+    | None ->
+        if not_throttled () then
+          debug "%s span does not exist or already finished" __FUNCTION__ ;
+        spans
+    | Some span_list -> (
+      match List.filter (fun x -> x.Span.context <> span.context) span_list with
+      | [] ->
+          SpanMap.remove key spans
+      | filtered_list ->
+          SpanMap.add key filtered_list spans
     )
 
-  let add_to_finished span =
-    Xapi_stdext_threads.Threadext.Mutex.execute lock (fun () ->
-        let spans, n = !finished_spans in
-        if n < Atomic.get max_spans then
-          finished_spans := (span :: spans, n + 1)
-        else if not_throttled () then
-          debug "%s exceeded max traces when adding to finished span table"
-            __FUNCTION__
-    )
+  let remove_from_spans span =
+    update_spans remove_from_spans_unlocked span ;
+    Some span
+
+  let rec add_to_finished span =
+    let ((spans, n) as old) = Atomic.get finished_spans in
+    if n < Atomic.get max_spans then
+      let next = (span :: spans, n + 1) in
+      if Atomic.compare_and_set finished_spans old next then
+        ()
+      else
+        (add_to_finished [@tailcall]) span
```

Comment:
Shouldn't there be a Thread.yield here, like in `Spans`?

Reply:
https://github.com/xapi-project/xen-api/pull/5958#discussion_r1762796073

We should eventually use Saturn (once it has the lock-free hashtable imported from Picos), or Kcas (which does these necessary yields with backoff).
I'll probably look at adding those to opam on the next OIL day, so we can start converting code to using them.